src/coreclr/vm/method.cpp

```diff
@@ -3035,8 +3035,8 @@ bool MethodDesc::DetermineAndSetIsEligibleForTieredCompilation()
         // Functional requirement - These methods have no IL that could be optimized
         !IsWrapperStub() &&
 
-        // Policy - Generating optimized code is not disabled
-        !IsJitOptimizationDisabledForSpecificMethod())
+        // Functions with NoOptimization or AggressiveOptimization don't participate in tiering
+        !IsJitOptimizationLevelRequested())
```

Comment:
I have a vague recollection being suggested that it is expensive to get the attributes and to avoid getting them eagerly, so the AggressiveOptimization check was avoided here and instead checked when the method is called. I forgot that the NoOptimization check is similar and was already being done. I wonder if it would be better to not check for NoOptimization here either.

An alternative would be to update `TieredCompilationManager::GetInitialOptimizationTier` to return the appropriate tier and `TieredCompilationManager::GetJitFlags` to set the appropriate JIT flags. Maybe NoOptimization could use the MinOptJitted tier.

I believe the tier 1 JIT flag was used for AggressiveOptimization in case tier 1 would opt to do more higher-cost optimizations in the future. It can be changed though.

Reply:
The changes themselves look ok to me since this is not adding additional startup cost, but I'm not sure if we should do this first and reevaluate later, or use the alternate solution that I mentioned. I don't know how expensive eager attribute lookup actually is. It would be possible to measure I guess. @noahfalk, thoughts?

Reply:
Short story: I think the PR as-is should be good.

Long story: There are two different cases for attributes that have different perf costs. Most attributes are encoded in the metadata custom attribute table which requires searching and is pretty slow. However the attributes here are pseudo-custom-attributes which are encoded as flags inline on the rows of other metadata tables and they can be accessed quickly. I think these MethodImpl bits are stored inline on the MethodDef record for example. While I don't know exactly what a call to MethodDesc::GetImplAttrs() costs, I'd hope it is single or maybe double digit ns. I believe it gets called several times during the load of each method.

Given that we were already calling MethodDesc::GetImplAttrs() inside the old call to IsJitOptimizationDisabledForSpecificMethod(), I think its fine to continue calling it the way this PR is doing it. For the purposes of this PR as long as we aren't regressing startup perf we should be good. Longer term we could also make performance improvements to startup but I'd propose driving that top down based profiling and identifying potential changes with the best cost/benefit. If the profiling identified GetImplAttrs() as a ripe target we could spend more time on it, and if not we'd spend our time on whatever else the profiling pointed us towards.

Reply:
Sounds good to me, thanks!

Reply:
Thanks for the clarification!